include/libhal-arm-mcu/stm32f1/adc.hpp

```diff
@@ -0,0 +1,157 @@
+   * @brief Defines the available adc peripherals that CAN be onboard the MCU.
+   * Having two adc's provides the capability to do dual-channel conversions, as
+   * well as do simultaneous sampling through dual-mode (not currently
+   * supported). Note that the XL-density stm32f1 MCU's only have adc1.
+   *
+   */
+  enum class adc_selection : hal::u8
+  {
+    adc1 = 0,
+    adc2 = 1,
+  };
+
+  /**
+   * @brief Construct a new adc peripheral manager object.
+   *
+   * @param p_adc_selection - The specified adc peripheral to use.
+   * @param p_lock - An externally declared lock to use for thread safety when
+   * trying to read from the adc's. This can be a basic_lock or any type that
+   * derives from it.
+   */
+  adc_peripheral_manager(adc_selection p_adc_selection,
+                         hal::basic_lock& p_lock);
+
+  /**
+   * @brief Creates and configures an adc channel under the calling adc
+   * peripheral manager.
+   *
+   * @param p_pin - The pin to be used for the channels analog input.
+   * @return channel - is an implementation of the `hal::adc` interface that
+   * represents a channel on the ADC peripheral.
+   */
+  channel acquire_channel(pins p_pin);
+
+  adc_peripheral_manager(adc_peripheral_manager const& p_other) = delete;
+  adc_peripheral_manager& operator=(adc_peripheral_manager const& p_other) =
+    delete;
+  adc_peripheral_manager(adc_peripheral_manager&& p_other) noexcept = delete;
+  adc_peripheral_manager& operator=(adc_peripheral_manager&& p_other) noexcept =
+    delete;
+
+private:
+  /**
+   * @brief Takes an analog input reading.
+   *
+   * @param p_pin - The pin to read from.
+   * @return float - The sampled adc value.
+   */
+  float read_channel(pins p_pin);
+
+  /// The lock to be used for thread safety with adc reads.
+  hal::basic_lock* m_lock;
+  /// A pointer to track the location of the registers for the specified adc
+  /// peripheral.
+  void* adc_reg_location;
+};
+
+/**
+ * @brief This class implements the `hal::adc` abstract base class. It creates
+ * channels to be used by the adc peripheral manager to read certain pins'
+ * analog input.
+ *
+ */
+class adc_peripheral_manager::channel : public hal::adc
+{
+public:
+  channel(channel const& p_other) = delete;
+  channel& operator=(channel const& p_other) = delete;
+  channel(channel&& p_other) noexcept = delete;
+  channel& operator=(channel&& p_other) noexcept = delete;
+
+private:
+  friend class adc_peripheral_manager;
+
+  // Constructor for channel. Needs to configure the given pin
+  /**
+   * @brief Constructs a channel object.
+   *
+   * @param p_manager - The adc peripheral manager that will be managing this
+   * channel resource.
+   * @param p_pin - The pin that will be used for this channels analog input.
+   */
+  channel(adc_peripheral_manager& p_manager,
+          adc_peripheral_manager::pins p_pin);
+
+  /**
+   * @brief Takes an analog input reading.
+   *
+   * @return float - The sampled adc value.
+   */
```

Comment:
```suggestion
```

You don't really need doxygen documentation for private things as they do not render in the doxygen API docs. Comments are useful to help explain something that isn't already obvious. And as for this function, the behavior is already mentioned in detail in the `hal::adc` so a caller would defer to that explanation rather than what you have here.